packages/test/src/test-integration-update.ts

```diff
@@ -0,0 +1,319 @@
+// provides test coverage for Update buffering: were it not for the buffering,
+// we would attempt to start performing the Update (validate and handle) before
+// its handler is set, since sdk-core sorts Update jobs with Signal jobs, i.e.
+// ahead of jobs such as startWorkflow and completeActivity that might result in
+// a setHandler call. Also note that we do need to make this guarantee to users,
+// because a user might know that their Update is in the first WFT, for example
+// because they are doing something similar to what this test does to achieve
+// that.
+
+// TODO: we currently lack a way to ensure, without race conditions, via SDK
+// APIs, that Updates are packaged together with startWorkflow in the first
+// Activation. In lieu of a non-racy implementation, the test below does the
+// following:
+// 1. Client sends and awaits startWorkflow.
+// 2. Client sends but does not await executeUpdate.
+// 3. Wait for long enough to be confident that the server handled the
+//    executeUpdate and is now waiting for the Update to advance to Completed.
+// 4. Start the Worker.
+
+const stateMutatingUpdate = wf.defineUpdate('stateMutatingUpdate');
+
+export async function setUpdateHandlerAndExit(): Promise<string> {
+  let state = 'initial';
+  const mutateState = () => void (state = 'mutated-by-update');
+  wf.setHandler(stateMutatingUpdate, mutateState);
+  // If an Update is present in the first WFT, then the handler should be called
+  // before the workflow exits and the workflow return value should reflect its
+  // side effects.
+  return state;
+}
+
+test('Update is always delivered', async (t) => {
+  const { createWorker, startWorkflow } = helpers(t);
+  const wfHandle = await startWorkflow(setUpdateHandlerAndExit);
+
+  // Race condition: wait long enough for the Update to have been admitted, so
+  // that it is in the first WFT, along with startWorkflow.
+  // eslint-disable-next-line @typescript-eslint/no-floating-promises
+  wfHandle.executeUpdate(stateMutatingUpdate);
+  await new Promise((res) => setTimeout(res, 1000));
+
+  const worker = await createWorker();
+  await worker.runUntil(async () => {
+    // Worker receives activation: [doUpdate, startWorkflow]
+    const wfResult = await wfHandle.result();
+    t.deepEqual(wfResult, 'mutated-by-update');
+  });
+});
+
+test('Two Updates in first WFT', async (t) => {
+  const { createWorker, startWorkflow } = helpers(t);
+  const wfHandle = await startWorkflow(workflowWithUpdates);
+
+  // Race condition: wait long enough for the Updates to have been admitted, so
+  // that they are in the first WFT, along with startWorkflow.
+  // eslint-disable-next-line @typescript-eslint/no-floating-promises
+  wfHandle.executeUpdate(update, { args: ['1'] });
+  // eslint-disable-next-line @typescript-eslint/no-floating-promises
+  wfHandle.executeUpdate(doneUpdate);
+  await new Promise((res) => setTimeout(res, 1000));
```

Comment:
Race condition will be fixed by https://github.com/dandavison/temporalio-sdk-typescript/pull/1